Windows/DSoundStream.cpp

```diff
@@ -232,6 +247,181 @@ void DSoundAudioBackend::Update() {
+
+WASAPIAudioBackend::WASAPIAudioBackend() : hThread_(NULL), sampleRate_(0), callback_(nullptr), threadData_(0) {
+	CoInitializeEx(NULL, COINIT_MULTITHREADED);
+}
+
+WASAPIAudioBackend::~WASAPIAudioBackend() {
+	CoUninitialize();
+	if (threadData_ == 0) {
+		threadData_ = 1;
+	}
+
+	if (hThread_ != NULL) {
+		WaitForSingleObject(hThread_, 1000);
+		CloseHandle(hThread_);
+		hThread_ = NULL;
+	}
+
+	if (threadData_ == 2) {
+		// blah.
+	}
+}
+
+unsigned int WINAPI WASAPIAudioBackend::soundThread(void *param) {
+	WASAPIAudioBackend *backend = (WASAPIAudioBackend *)param;
+	return backend->RunThread();
+}
+
+bool WASAPIAudioBackend::Init(HWND window, StreamCallback callback, int sampleRate) {
+	threadData_ = 0;
+	callback_ = callback;
+	sampleRate_ = sampleRate;
+	hThread_ = (HANDLE)_beginthreadex(0, 0, soundThread, (void *)this, 0, 0);
+	SetThreadPriority(hThread_, THREAD_PRIORITY_ABOVE_NORMAL);
+	return true;
+}
+
+int WASAPIAudioBackend::RunThread() {
+	IMMDeviceEnumerator *pDeviceEnumerator;
+	IMMDevice *pDevice;
+	IAudioClient *pAudioInterface;
+	IAudioRenderClient *pAudioRenderClient;
+	WAVEFORMATEXTENSIBLE *pDeviceFormat;
+	DWORD flags = 0;
+	REFERENCE_TIME hnsBufferDuration, hnsActualDuration;
+	UINT32 pNumBufferFrames;
+	UINT32 pNumPaddingFrames, pNumAvFrames;
+	hnsBufferDuration = REFTIMES_PER_SEC;
+
+	HRESULT hresult;
+	hresult = CoCreateInstance(CLSID_MMDeviceEnumerator,
+		NULL, /*Object is not created as the part of the aggregate */
+		CLSCTX_ALL, IID_IMMDeviceEnumerator, (void**)&pDeviceEnumerator);
+
+	hresult = pDeviceEnumerator->GetDefaultAudioEndpoint(eRender, eMultimedia, &pDevice);
+	hresult = pDevice->Activate(IID_IAudioClient, CLSCTX_ALL, NULL, (void**)&pAudioInterface);
+	hresult = pAudioInterface->GetMixFormat((WAVEFORMATEX**)&pDeviceFormat);
+	hresult = pAudioInterface->Initialize(AUDCLNT_SHAREMODE_SHARED, 0, hnsBufferDuration, 0, &pDeviceFormat->Format, NULL);
+	hresult = pAudioInterface->GetService(IID_IAudioRenderClient, (void**)&pAudioRenderClient);
+	hresult = pAudioInterface->GetBufferSize(&pNumBufferFrames);
+
+	sampleRate_ = pDeviceFormat->Format.nSamplesPerSec;
+
+	if (pDeviceFormat->Format.wFormatTag == 0xFFFE) {
+		if (!memcmp(&pDeviceFormat->SubFormat, &KSDATAFORMAT_SUBTYPE_IEEE_FLOAT, sizeof(pDeviceFormat->SubFormat))) {
+			// printf("float format\n");
+		} else {
+			ERROR_LOG(SCEAUDIO, "Got unexpected WASAPI stream format, expected float!");
+		}
+	}
+
+	short *shortBuf = new short[pNumBufferFrames * pDeviceFormat->Format.nChannels];
+		
+	BYTE *pData;
+	hresult = pAudioRenderClient->GetBuffer(pNumBufferFrames, &pData);
+
+	int numFloats = pNumBufferFrames * pDeviceFormat->Format.nChannels;
+	float *ptr = (float *)pData;
+	for (int i = 0; i < numFloats; i++) {
+		ptr[i] = 0.0f;
+	}
+
+	hresult = pAudioRenderClient->ReleaseBuffer(pNumBufferFrames, flags);
+	hnsActualDuration = (double)REFTIMES_PER_SEC * pNumBufferFrames / pDeviceFormat->Format.nSamplesPerSec;
```

Comment:
This causes a warning, would it be better to just keep it an integer (u64 ought to be enough?)

-[Unknown]